Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -113,19 +132,40 @@ extension HTTP1ToRawGRPCServerCodec: ChannelInboundHandler {
 extension HTTP1ToRawGRPCServerCodec: ChannelOutboundHandler {
   public typealias OutboundIn = RawGRPCServerResponsePart
   public typealias OutboundOut = HTTPServerResponsePart
-  
+
   public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
     let responsePart = self.unwrapOutboundIn(data)
     switch responsePart {
-    case .headers(let headers):
-      //! FIXME: Should return a different version if we want to support pPRC.
-      ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: .init(major: 2, minor: 0), status: .ok, headers: headers))), promise: promise)
+    case .headers(var headers):
+      let version: HTTPVersion
+      if let contentType = contentType {
+        switch contentType {
+        case .text, .web:
+          headers.add(name: "content-type", value: contentType.rawValue)
+          version = .init(major: 1, minor: 1)
+        case .binary:
+          headers.add(name: "Content-Type", value: contentType.rawValue)
+          version = .init(major: 2, minor: 0)
+        }
+      } else {
+        version = .init(major: 2, minor: 0)
+      }
+
+      ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: version, status: .ok, headers: headers))), promise: promise)
     case .message(var messageBytes):
       // Write out a length-delimited message payload. See `channelRead` fpor the corresponding format.
       var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.readableBytes + 5)
       responseBuffer.write(integer: Int8(0))  // Compression flag: no compression
       responseBuffer.write(integer: UInt32(messageBytes.readableBytes))
       responseBuffer.write(buffer: &messageBytes)
+
+      if contentType == .text {
+        if let binaryData = responseBuffer.readData(length: responseBuffer.readableBytes) {
+          let encodedData = binaryData.base64EncodedString()
+          responseBuffer = ctx.channel.allocator.buffer(capacity: encodedData.utf8.count)
+          responseBuffer.write(string: encodedData)
+        }
+      }
```

Comment:
You're correct. Had to process the messages in a different way, since to have the base64 encoding properly aligned, I had to receive all of the streaming messages plus the trailers in order to aggregate them together and calculate the base64 of the complete response.

I realize that the complexity of HTTP1ToRawGRPC is getting out of hand, so after this is merged, the first thing I'll do is extract the grpcweb base64 wrapping into a different handler and only add it for the HTTP1.1 pipeline. I've added a todo to do that.

Reply:
I like the idea of doing the base64 decoding/encoding in a separate handler; that's a great way to abstract this away.